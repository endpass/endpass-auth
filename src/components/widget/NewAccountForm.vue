<template>
  <form @submit.prevent="handleSubmit">
    <widget-input
      v-model="password"
      v-validate="'required|min:8'"
      :error="errors.first('password')"
      :disabled="isLoading"
      :placeholder="$t('components.passwordForm.enterWalletPassword')"
      data-test="new-account-password-input"
      type="password"
      name="password"
    />
    <option-button
      :is-big="true"
      type="submit"
      data-test="new-account-submit-button"
      :disabled="!isFormValid || isLoading"
    >
      {{ submitButtonLabel }}
    </option-button>
    <option-button
      :is-big="true"
      :disabled="isLoading"
      data-test="new-account-cancel-button"
      @click="handleCancelClick"
    >
      {{ $t('components.widgetNewAccount.cancel') }}
    </option-button>
  </form>
</template>

<script>
import { mapActions } from 'vuex';
import formMixin from '@/mixins/form';
import WidgetInput from './WidgetInput.vue';
import OptionButton from './OptionButton.vue';

export default {
  name: 'WidgetNewAccountForm',

  props: {
    currentAccount: {
      type: String,
      default: null,
    },

    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    password: '',
  }),

  computed: {
    submitButtonLabel() {
      if (this.isLoading) {
        return this.$t('components.widgetNewAccount.creatingAccount');
      }

      return this.$t('components.widgetNewAccount.createAccount');
    },
  },

  watch: {
    password() {
      const { errors } = this;

      if (errors.firstById('incorrectPassword')) {
        errors.removeById('incorrectPassword');
      }
    },
  },

  methods: {
    ...mapActions(['createAccountFromWidget']),

    async handleSubmit() {
      if (!this.isFormValid) return;

      try {
        await this.createAccountFromWidget({
          address: this.currentAccount,
          password: this.password,
        });
        this.$emit('cancel');
      } catch (err) {
        this.errors.add({
          id: 'incorrectPassword',
          field: 'password',
          msg: err.message,
        });
      }
    },

    handleCancelClick() {
      this.$emit('cancel');
    },
  },

  mixins: [formMixin],

  components: {
    WidgetInput,
    OptionButton,
  },
};
</script>
