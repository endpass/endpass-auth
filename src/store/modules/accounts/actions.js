import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';
import isV3 from '@endpass/utils/isV3';
import mapToQueryString from '@endpass/utils/mapToQueryString';
import signerService from '@/service/signer';
import identityService from '@/service/identity';
import permissionsService from '@/service/permissions';
import settingsService from '@/service/settings';
import modeService from '@/service/mode';
import cryptoDataService from '@/service/cryptoData';
import bridgeMessenger from '@/class/singleton/bridgeMessenger';

import {
  accountChannel,
  authChannel,
  permissionChannel,
} from '@/class/singleton/channels';
import { Answer } from '@/class';
import { METHODS, ORIGIN_HOST } from '@/constants';

const auth = async ({ state, dispatch }, { email, serverMode }) => {
  const { type, serverUrl } = serverMode;
  const redirectUrl = get(state, 'authParams.redirectUrl', '');
  const queryParamsToAppend = {
    mode: type,
  };

  if (serverUrl) {
    Object.assign(queryParamsToAppend, {
      serverUrl,
    });
  }

  const request = identityService.auth(
    email,
    mapToQueryString(redirectUrl, queryParamsToAppend),
  );

  await dispatch('handleAuthRequest', {
    request,
    email,
    link: true,
  });
};

const authWithGoogle = async ({ dispatch }, { email, idToken }) => {
  const request = identityService.authWithGoogle(idToken);

  await dispatch('handleAuthRequest', {
    request,
    email,
  });
};

const authWithGitHub = async ({ commit }, code) => {
  commit('changeLoadingStatus', true);

  try {
    const res = await identityService.authWithGitHub(code);

    if (!res.success) throw new Error(res.message || 'Auth failed!');

    settingsService.clearLocalSettings();

    const type = get(res, 'challenge.challengeType');
    if (type === 'otp') {
      commit('setOtpEmail', res.email);
    }
  } catch (err) {
    console.error(err);

    throw new Error(err.message);
  } finally {
    commit('changeLoadingStatus', false);
  }
};

const authWithHydra = async ({ commit }, { challengeId, password }) => {
  commit('changeLoadingStatus', true);
  let res;

  try {
    const { email, v3Keystore } = await permissionsService.getLoginDetails();

    const { signature } = await signerService.signDataWithAccount({
      account: v3Keystore,
      data: email,
      password,
    });

    res = await permissionsService.login({
      challengeId,
      signature,
    });
  } catch (err) {
    console.error(err);
    throw new Error('Password is incorrect');
  } finally {
    commit('changeLoadingStatus', false);
  }
  return res;
};

const checkHydraLoginRequirements = async ({ commit }, challengeId) => {
  commit('changeLoadingStatus', true);

  try {
    const res = await permissionsService.getLoginSkipStatus(challengeId);

    return res;
  } catch (err) {
    throw new Error('Checking Hydra login status');
  } finally {
    commit('changeLoadingStatus', false);
  }
};

const getConsentDetails = (ctx, consentChallenge) =>
  permissionsService.getConsentDetails(consentChallenge);

const grantPermissionsWithHydra = (ctx, { consentChallenge, scopesList }) =>
  permissionsService.grantPermissions({ consentChallenge, scopesList });

const handleAuthRequest = async ({ commit }, { email, request, link }) => {
  commit('changeLoadingStatus', true);

  try {
    const res = await request;

    if (!res.success) throw new Error('Auth failed!');

    settingsService.clearLocalSettings();

    const type = get(res, 'challenge.challengeType');
    if (type === 'otp') {
      commit('setOtpEmail', email);
    } else if (link) {
      commit('setSentStatus', true);
    }
  } catch (err) {
    throw err;
  } finally {
    commit('changeLoadingStatus', false);
  }
};

const confirmAuthViaOtp = async ({ commit }, { email, code }) => {
  commit('changeLoadingStatus', true);

  try {
    await identityService.otpAuth(email, code);
  } catch (err) {
    throw err;
  } finally {
    commit('changeLoadingStatus', false);
  }
};

const confirmAuth = (ctx, serverMode) => {
  authChannel.put(Answer.createOk(serverMode));
};

const cancelAuth = () => {
  authChannel.put(Answer.createFail('Auth was canceled by user!'));
};

const getSettings = async ({ dispatch }) => {
  const settings = await identityService.getSettings();
  const { lastActiveAccount } = settings;
  let account = null;

  if (lastActiveAccount) {
    account = await dispatch('getAccount', lastActiveAccount);
  }

  if (!lastActiveAccount || !isV3(account)) {
    const lastAccount = await dispatch('getFirstPrivateAccount');

    Object.assign(settings, {
      lastActiveAccount: get(lastAccount, 'address'),
    });
  }

  return settings;
};

const defineSettings = async ({ state, dispatch, commit, getters }) => {
  const { demoData } = getters;
  const settings = demoData ? state.settings : await dispatch('getSettings');
  const mergedSettings = settingsService.mergeSettings(settings);

  settingsService.setLocalSettings(mergedSettings);

  const newSettings = {
    ...settings,
    ...mergedSettings,
  };

  commit('setSettings', newSettings);
};

const setSettings = async (ctx, payload) => {
  settingsService.setLocalSettings(payload);
};

const updateSettings = async ({ state, commit, dispatch }, payload) => {
  commit('changeLoadingStatus', true);

  try {
    await dispatch('setSettings', payload);
    await dispatch('defineSettings');

    const { settings } = state;
    const settingsToSend = {
      activeAccount: settings.lastActiveAccount,
      activeNet: settings.net,
    };
    const answer = Answer.createOk({
      type: 'update',
      settings: settingsToSend,
    });

    commit('setBalance', null);

    bridgeMessenger.send(METHODS.CHANGE_SETTINGS_REQUEST, settingsToSend);
    accountChannel.put(answer);
  } catch (err) {
    throw new Error('Something went wrong, try again later');
  } finally {
    commit('changeLoadingStatus', false);
  }
};

const checkAccountExists = () => identityService.checkAccountExist();

const getAccounts = async ({ commit }) => {
  // TODO: check `getAccounts` usages
  try {
    const res = await identityService.getAccounts();
    const accounts = await Promise.all(
      res.map(address => identityService.getAccountInfo(address)),
    );

    commit(
      'setAccounts',
      accounts.filter(account => !/^xpub/.test(account.address)),
    );
    commit('setAuthStatus', true);
  } catch (err) {
    commit('setAccounts', null);
    commit('setAuthStatus', false);
  }
};

const defineOnlyV3Accounts = async ({ commit, getters }) => {
  if (getters.demoData) {
    commit('setAuthStatus', true);
    return;
  }

  try {
    const res = await identityService.getAccounts();

    const accounts = await Promise.all(
      res
        .filter(address => !/^xpub/.test(address))
        .map(address => identityService.getAccountWithInfo(address)),
    );

    commit(
      'setAccounts',
      accounts.filter(account => isV3(account)).map(({ info }) => info),
    );
    commit('setAuthStatus', true);
  } catch (err) {
    commit('setAccounts', null);
    commit('setAuthStatus', false);
  }
};

const getAccount = async (ctx, address) => {
  const res = await identityService.getAccount(address);

  return res;
};

const getFirstPrivateAccount = async ({ state, dispatch }) => {
  if (isEmpty(state.accounts)) {
    await dispatch('defineOnlyV3Accounts');
  }

  const { accounts } = state;

  return !accounts
    ? null
    : accounts.find(account => account.type !== 'PublicAccount') || null;
};

const waitAccountCreate = async ({ commit }) => {
  const res = await identityService.waitAccountCreate();

  commit('setAccounts', res);
};

const openCreateAccountPage = async () => {
  window.open(`${ENV.VUE_APP_WALLET_URL}?closeAfterCreateWallet=true`);
};

const closeAccount = async () => {
  accountChannel.put(Answer.createOk({ type: 'close' }));
};

const getRecoveryIdentifier = async ({ state, commit }) => {
  commit('changeLoadingStatus', true);

  try {
    const identifier = await identityService.getRecoveryIdentifier(
      state.otpEmail,
    );

    commit('setRecoveryIdentifier', identifier);
  } catch (err) {
    throw err;
  } finally {
    commit('changeLoadingStatus', false);
  }
};

const recover = async ({ state, commit }, { seedPhrase }) => {
  commit('changeLoadingStatus', true);

  try {
    const signature = await signerService.recover({
      seedPhrase,
      recoveryIdentifier: state.recoveryIdentifier,
    });

    const { success } = await identityService.recover(
      state.otpEmail,
      signature,
      state.authParams.redirectUrl,
    );

    commit('setSentStatus', true);

    return success;
  } catch (err) {
    throw err;
  } finally {
    commit('changeLoadingStatus', false);
  }
};

const validateCustomServer = (ctx, serverUrl) =>
  modeService.validateIdentityServer(serverUrl);

const setupDemoData = async ({ commit }, demoData) => {
  if (!demoData || !demoData.v3KeyStore) {
    return;
  }

  const { address } = demoData.v3KeyStore;
  const accounts = [{ address }];

  commit('setDemoData', demoData);
  commit('setAccounts', accounts);
  commit('setSettings', {
    lastActiveAccount: address,
    net: demoData.activeNet,
  });
};

const signPermission = async (store, { password }) => {
  const res = await identityService.getAuthPermission();

  if (res.success === false) {
    throw new Error(res.message);
  }
  const signature = await signerService.getSignedRequest({
    v3KeyStore: res.keystore,
    password,
    request: {
      params: [ORIGIN_HOST],
    },
  });
  await identityService.setAuthPermission(signature);
  permissionChannel.put(Answer.createOk());
};

const cancelSignPermission = () => {
  permissionChannel.put(Answer.createFail());
};

const waitLogin = async ({ dispatch }) => {
  await identityService.waitLogin();
  await dispatch('defineAuthStatus');
  // authChannel.put(Answer.createOk());
};

const defineAuthStatus = async ({ commit }) => {
  const status = await identityService.getAuthStatus();
  commit('setAuthByCode', status);
  return status;
};

const getAccountBalance = async (ctx, { address, net }) => {
  const { balance } = await cryptoDataService.getAccountBalance({
    network: net,
    address,
  });

  return balance;
};

const subscribeOnBalanceUpdates = ({ state, commit, dispatch }) => {
  const handler = () =>
    setTimeout(async () => {
      const address = get(state.settings, 'lastActiveAccount');
      const net = get(state.settings, 'net', 1);

      if (address) {
        try {
          const balance = await dispatch('getAccountBalance', { address, net });

          commit('setBalance', balance);
        } catch (err) {
          commit('setBalance', null);
        }
      }

      handler();
    }, 1500);

  handler();
};

export default {
  auth,
  authWithGoogle,
  authWithGitHub,
  authWithHydra,
  checkAccountExists,
  grantPermissionsWithHydra,
  cancelAuth,
  confirmAuth,
  confirmAuthViaOtp,
  handleAuthRequest,
  defineSettings,
  getAccount,
  getAccounts,
  getFirstPrivateAccount,
  getSettings,
  defineOnlyV3Accounts,
  openCreateAccountPage,
  waitAccountCreate,
  waitLogin,
  setSettings,
  updateSettings,
  closeAccount,
  getRecoveryIdentifier,
  recover,
  validateCustomServer,
  defineAuthStatus,
  signPermission,
  cancelSignPermission,
  setupDemoData,
  getConsentDetails,
  getAccountBalance,
  subscribeOnBalanceUpdates,
  checkHydraLoginRequirements,
};
