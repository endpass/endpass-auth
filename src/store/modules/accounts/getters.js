export default {
  demoData: state => state.demoData,
  isAuthorized: state => state.isLogin && state.isPermission,
};
