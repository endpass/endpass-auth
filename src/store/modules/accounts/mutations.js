import isEmpty from 'lodash/isEmpty';

const setAuthParams = (state, params) => {
  state.authParams = params;
};

const setAuthStatus = (state, flag) => {
  // TODO: merge with setAuthByCode functionality
  state.isLogin = flag;
  state.isPermission = flag;
};

const setAuthByCode = (state, code) => {
  state.isLogin = code !== 401;
  state.isPermission = !(code === 403 || code === 401);
};

const setOtpEmail = (state, email) => {
  state.otpEmail = email;
};

const setSentStatus = (state, status) => {
  state.linkSent = status;
};

const setAccounts = (state, accounts) => {
  state.accounts = isEmpty(accounts) ? [] : [...accounts];
};

const setSettings = (state, settings) => {
  state.settings = settings;
};

const setRecoveryIdentifier = (state, recoveryIdentifier) => {
  state.recoveryIdentifier = recoveryIdentifier;
};

const setDemoData = (state, demoData) => {
  state.demoData = demoData;
};

const logout = state => {
  state.otpEmail = null;
  state.authParams = null;
  state.isPermission = false;
  state.isLogin = false;
  state.accounts = [];
  state.settings = null;
  state.recoveryIdentifier = null;
};

export default {
  logout,
  setDemoData,
  setAuthParams,
  setAuthStatus,
  setOtpEmail,
  setAccounts,
  setSentStatus,
  setSettings,
  setRecoveryIdentifier,
  setAuthByCode,
};
