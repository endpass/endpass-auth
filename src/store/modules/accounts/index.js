import actions from './actions';
import mutations from './mutations';
import getters from './getters';

const state = {
  linkSent: false,
  authParams: null,
  otpEmail: null,
  accounts: [],
  settings: null,
  recoveryIdentifier: null,
  isAuthorized: false,
};

export default {
  state,
  actions,
  mutations,
  getters,
};
