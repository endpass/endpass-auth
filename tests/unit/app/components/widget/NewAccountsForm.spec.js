import Vuex from 'vuex';
import validation from '@/validation';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import NewAccountForm from '@/components/widget/NewAccountForm.vue';
import setupI18n from '@/locales/i18nSetup';
import { address } from '@unitFixtures/accounts';
import VeeValidate from 'vee-validate';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VeeValidate);
localVue.use(validation);
const i18n = setupI18n(localVue);

jest.useFakeTimers();

describe('NewAccountForm', () => {
  let wrapper;
  let store;
  let storeData;
  let widgetModule;

  beforeEach(() => {
    jest.clearAllMocks();
    widgetModule = {
      actions: {
        createAccountFromWidget: jest.fn(),
      },
    };
    storeData = {
      modules: {
        widget: widgetModule,
      },
    };

    store = new Vuex.Store(storeData);
    wrapper = shallowMount(NewAccountForm, {
      localVue,
      store,
      i18n,
      sync: false,
      provide: {
        theme: 'default',
      },
    });
  });

  describe('render', () => {
    it('should correctly render NewAccountForm component', () => {
      expect(wrapper.name()).toBe('WidgetNewAccountForm');
      expect(wrapper.html()).toMatchSnapshot();
      expect(wrapper.find('[data-test=new-account-submit-button]').text()).toBe(
        i18n.t('components.widgetNewAccount.createAccount'),
      );
    });

    it('should change submit button label on loading', async () => {
      expect.assertions(1);

      wrapper.setProps({
        isLoading: true,
      });

      await wrapper.vm.$nextTick();

      expect(wrapper.find('[data-test=new-account-submit-button]').text()).toBe(
        i18n.t('components.widgetNewAccount.creatingAccount'),
      );
    });
  });

  describe('behavior', () => {
    it('should emit cancel event on cancel button click', () => {
      wrapper.find('[data-test=new-account-cancel-button]').vm.$emit('click');

      expect(wrapper.emitted().cancel).not.toBeFalsy();
    });

    it('should submit password and current account on form submit', async () => {
      expect.assertions(1);

      const password = '12341234';

      wrapper.setProps({
        currentAccount: address,
      });
      wrapper
        .find('[data-test=new-account-password-input]')
        .vm.$emit('input', password);

      await wrapper.vm.$nextTick();
      await wrapper.vm.$nextTick();

      wrapper.find('form').trigger('submit');

      await global.flushPromises();

      expect(widgetModule.actions.createAccountFromWidget).toBeCalledWith(
        expect.any(Object),
        {
          password,
          address,
        },
        undefined,
      );
    });

    it('should set error if submit failed', async () => {
      const error = new Error('foo');

      widgetModule.actions.createAccountFromWidget.mockRejectedValueOnce(error);

      const password = '12341234';

      wrapper.setProps({
        currentAccount: address,
      });
      wrapper
        .find('[data-test=new-account-password-input]')
        .vm.$emit('input', password);

      await wrapper.vm.$nextTick();
      await wrapper.vm.$nextTick();

      wrapper.find('form').trigger('submit');

      await global.flushPromises();

      expect(
        wrapper.find('[data-test=new-account-password-input]').attributes()
          .error,
      ).toBe(error.message);
    });
  });
});
