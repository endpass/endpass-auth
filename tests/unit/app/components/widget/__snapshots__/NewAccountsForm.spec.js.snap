// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAccountForm render should correctly render NewAccountForm component 1`] = `
<form>
  <widget-input-stub value="" placeholder="Enter your password..." data-test="new-account-password-input" type="password" name="password"></widget-input-stub>
  <option-button-stub isbig="true" type="submit" data-test="new-account-submit-button" disabled="true">
    Create account
  </option-button-stub>
  <option-button-stub isbig="true" data-test="new-account-cancel-button">
    Cancel
  </option-button-stub>
</form>
`;
